src/core/geometry/qgsgeos.cpp

```diff
@@ -540,6 +541,44 @@ QgsAbstractGeometry *QgsGeos::symDifference( const QgsAbstractGeometry *geom, QS
   return overlay( geom, OverlaySymDifference, errorMsg, parameters ).release();
 }
 
+static bool isZVerticalLine( const QgsAbstractGeometry *geom, double tolerance = 4 * std::numeric_limits<double>::epsilon() )
+{
+  // checks if the Geometry if a purely vertical 3D line LineString Z((X Y Z1, X Y Z2, ..., X Y Zn))
+  // This is needed because QgsGeos is not able to handle this type of geometry on distance computation.
+
+  if ( geom->wkbType() != Qgis::WkbType::LineStringZ && geom->wkbType() != Qgis::WkbType::LineStringZM )
+  {
+    return false;
+  }
+
+  std::unique_ptr<QgsLineString> line( qgsgeometry_cast<QgsLineString *>( geom->clone() ) );
+  bool isVertical = true;
+  if ( line )
+  {
+    const int nrPoints = line->numPoints();
+    if ( nrPoints == 1 )
+    {
+      return true;
+    }
+
+    // if the 2D part of two points of the line are different, this means
+    // that the line is not purely vertical
+    const double sqrTolerance = tolerance * tolerance;
+    for ( int iVert = nrPoints - 1, jVert = 0; jVert < nrPoints; iVert = jVert++ )
+    {
+      const QgsPoint pi = geom->vertexAt( QgsVertexId( 0, 0, iVert ) );
```

Comment:
vertexAt is a slow approach to iterate through vertices -- better is something like https://github.com/qgis/QGIS/blob/master/src/core/symbology/qgssymbollayerutils.cpp#L4452 .

Reply:
This should be good now.